### PR TITLE
Architecture/viper/search scene

### PR DESCRIPTION
**Pull request by matej494**

### CODE REVIEW

File: FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+//
+
+import Foundation
+
+class SearchDataSource: DataSourceProtocol {
+    var sections = [SearchSection]()
+    private var locations = [Location]()
+    private var filterText: String?
+    
+    func addLocations(_ locations: [Location]) {
```
Comment (dkulundzic) [inline FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift:18]: As a convention, I would always extract `public`/`internal` methods or stored/computed properties into a separate extension of the host type.

In example, `addLocations` and `filterLocations` would be extracted into a `public` or `internal`(depending on the requirement) extension.

As a result, we get increased readability and understanding of the type's `API`.

File: FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        locations.reversed().forEach { location in
+            if !self.locations.contains(where: { $0 == location }) {
+                self.locations.insert(location, at: 0)
+            }
+        }
+        buildSections(forText: filterText)
+    }
+    
+    func filterLocations(forText text: String) {
```
Comment (dkulundzic) [inline FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift:27]: Same as above.

File: FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        buildSections(forText: text)
+    }
+}
+
+private extension SearchDataSource {
+    func buildSections(forText text: String?) {
+        sections.removeAll()
+        var rows = [SearchRow]()
+        if let text = text,
+            !text.isEmpty {
```
Comment (dkulundzic) [inline FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift]: What do you think about turning

```
if let text = text,
	            !text.isEmpty
```

into 

`if let text = text, !text.isEmpty`

Much more readable.

File: FactoryWeather/App/VIP Search/SearchInteractor.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+class SearchInteractor {
+    var presenter: SearchPresentationLogic?
+    lazy var locationsWorker = LocationsWorker()
+}
+
+// MARK: - Business Logic
+extension SearchInteractor: SearchBusinessLogic {
+    func getLocations(forText text: String) {
+        locationsWorker.getLocations(forText: text)
+            .then { [weak self] locations in self?.presenter?.presentLocations(locations) }
```
Comment (dkulundzic) [inline FactoryWeather/App/VIP Search/SearchInteractor.swift]: We can shorten this by using anonymous parameters:
`.then { [weak self] in self?.presenter?.presentLocations($0) }`

File: FactoryWeather/App/VIP Search/SearchInteractor.swift [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            .catch { [weak self] error in self?.presenter?.presentError(error) }
```
Comment (dkulundzic) [inline FactoryWeather/App/VIP Search/SearchInteractor.swift]: We can shorten this by using anonymous parameters:
`.catch { [weak self] error in self?.presenter?.presentError(error) }`

File: FactoryWeather/Common/Network/Geo Names/GeoNamesApiManager.swift [hunk 6]
```diff
@@ -7,28 +7,27 @@
...
-            do {
-                guard let unwrappedData = data
-                    else { return DispatchQueue.main.async { failure(NetworkManagerError.dataUnwrappingFailure) } }
-                let geoName = try JSONDecoder().decode(GeoName.self, from: unwrappedData)
-                return DispatchQueue.main.async { success(geoName.locations) }
-            } catch {
-                return DispatchQueue.main.async { failure(NetworkManagerError.parsingDataFailure) }
-            }
+    static func getLocations(forText text: String) -> Promise<[Location]> {
+        return Promise(on: DispatchQueue.global(qos: .background)) { fulfill, reject in
```
Comment (dkulundzic) [inline FactoryWeather/Common/Network/Geo Names/GeoNamesApiManager.swift:14]: It's actually not necessary to publish this `Promise` on a `background` queue, as `URLSession.shared.dataTask` will dispatch on a `non-main` queue by `default`. Regardless, it's a good idea to set it as a precaution.

File: FactoryWeather/App/VIP Search/SearchViewController.swift [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+protocol SearchDisplayLogic: class {
+    func displayLocations(_ locations: [Location])
+    func displayError(_ error: Error)
+}
+
+class SearchViewController: UIViewController {
+    var interactor: SearchBusinessLogic?
+    var router: SearchRoutingLogic?
+    private let contentView: SearchContentView
+    private let activityIndicatorView = UIActivityIndicatorView.autolayoutView()
```
Comment (dkulundzic) [inline FactoryWeather/App/VIP Search/SearchViewController.swift:22]: Just a note, `activityIndicatorView` should not be located in the `view controller`, but the `content view`.

**Review by dkulundzic — CHANGES_REQUESTED** [review]
Architecturally, this is spot on, well done. I tried nitpicking, but couldn't find anything 🙂 🚀 

There are minor comments about the code though, nothing big.

File: FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift (hunk 3 above)
Comment (matej494) [reply FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift]: It's a lot of `text`s in one row, but I'll change that. I got used to this way of writing statements so it's more readable to me.

Comment (dkulundzic) [reply FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift]: I don't mind if you leave it as is. 

My personal preference is that I find short `if`'s like this more readable as a one-liner 🙂

File: FactoryWeather/Common/Network/Geo Names/GeoNamesApiManager.swift (hunk 6 above)
Comment (matej494) [reply FactoryWeather/Common/Network/Geo Names/GeoNamesApiManager.swift:14]: Ok, I used it just to try it out.

File: FactoryWeather/App/VIP Search/SearchViewController.swift (hunk 7 above)
Comment (matej494) [reply FactoryWeather/App/VIP Search/SearchViewController.swift:22]: Yep, we discussed that on another PR, but agreed to leave it as is. (This is leftover from old code and could be improved, but you said the main task of this sprint is to learn the architecture)

Comment (dkulundzic) [reply FactoryWeather/App/VIP Search/SearchViewController.swift:22]: Roger.

File: FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift (hunk 1 above)
Comment (matej494) [reply FactoryWeather/App/VIP Search/DataSource/SearchDataSource.swift:18]: Can I write just `extension SearchDataSource` or should it be `internal extension SearchDataSource`?

**Review by dkulundzic — APPROVED** [review]
Lift off! 🚀 🚀

**Pull request merged**